mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,328 @@
+            exterior=boundaries[btag].exterior_grad_q(discr, btag=btag, t=t,
+                                                      grad_q=r, eos=eos))
+        r_flux_db = r_flux_db + _facial_flux_r(discr, r_tpair=r_tpair)
+    # Total grad(Q) flux element boundaries
+    r_flux_bnd = r_flux_int + r_flux_pb + r_flux_db
+
+    # Return the AV RHS term
+    return discr.inverse_mass(-div_r_vol + discr.face_mass(r_flux_bnd))
+
+
+def artificial_viscosity(discr, t, eos, boundaries, q, alpha, **kwargs):
+    """Interface :function:`av_operator` with backwards-compatible API."""
+    from warnings import warn
+    warn("Do not call artificial_viscosity; it is now called av_operator. This"
+         "function will disappear in 2021", DeprecationWarning, stacklevel=2)
+    return av_operator(discr=discr, eos=eos, boundaries=boundaries,
+                       q=q, alpha=alpha, t=t)
+
+
+def compute_smoothness_indicator():
+    """Compute the smoothness indicator for all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j,k]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in and
+        0<=k<ndiscr_nodes_in}""",
+        "result[iel,idof] = "
+        "sum(k,vec[iel,k]*vec[iel,k]*modes[k])/sum(j,"
+        " vec[iel,j]*vec[iel,j]+1.0e-12/ndiscr_nodes_in)",
+        name="smooth_comp",
+    )
+    return knl
+
+
+def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
+    r"""Calculate the smoothness indicator.
+
+    Parameters
+    ----------
+    u: meshmode.dof_array.DOFArray
+        The field that is used to calculate the smoothness indicator.
+
+    kappa
+        An optional argument that controls the width of the transition from 0 to 1.
+    s0
+        An optional argument that sets the smoothness level to limit
+        on. Values in the range $(-\infty,0]$ are allowed, where $-\infty$ results in
+        all cells being tagged and 0 results in none.
+
+    Returns
+    -------
+    meshmode.dof_array.DOFArray
+        The elementwise constant values between 0 and 1 which indicate the smoothness
+        of a given element.
+    """
+    assert isinstance(u, DOFArray)
+
+    def get_indicator():
+        return compute_smoothness_indicator()
+
+    # Convert to modal solution representation
+    modal_map = discr.connection_from_dds(DD_VOLUME, DD_VOLUME_MODAL)
+    uhat = modal_map(u)
+
+    # Compute smoothness indicator value
+    actx = u.array_context
+    indicator = discr.empty(actx, dtype=u.entry_dtype)
+    for group in discr.discr_from_dd("vol").groups:
+        mode_ids = group.mode_ids()
+        modes = len(mode_ids)
+        order = group.order
+        highest_mode = np.ones(modes)
+        for mode_index, mode_id in enumerate(mode_ids):
+            highest_mode[mode_index] = highest_mode[mode_index] * (
+                sum(mode_id) == order
+            )
```

Comment:
This is probably more readable when done in a list comprehension.
`[1 if sum(mode_id) == order else 0 for mode_id in mode_ids]` or so.